### PR TITLE
making data preprocessing step configurable with two options no preprocessing and feature type split

### PR DESCRIPTION
**Pull request by rabsr**
New approach for #900 : Data preprocessing step of AutoML system will have two options:

- use feature_type split method (existing implementation)
- disable step by selecting no_preprocessing

Introduced new parameters:

- include_data_preprocessors
- exclude_data_preprocessors

If none of the parameter is set by user, by default, it will add 'no_preprocessing' to step exclude_data_preprocessors and will only use existing FeatureTypeSplit component.

### CODE REVIEW

**Comment by mfeurer** [comment]
Hi @rabsr, I finally got around looking at this PR. Thanks again for all your work!

Let me briefly check whether I understand this PR correctly: are you proposing to add the possibility to drop the whole data preprocessing pipeline? And could you briefly state the goal of this - is it 1) being able to configure whether the data preprocessing should be used or 2) being able to disable the data preprocessing at all?

**Comment by rabsr** [comment]
@mfeurer The plan is to make data preprocessing step configurable. Such that each step within data preprocessing pipeline is also configurable. So user can even configure whether he wants to perform rescaling or categorical encoding etc.  This will provide more control of pipeline configuration space and also reduce no. of hyperparameters to tune as per requirement.

The idea of this PR is to make data preprocessing step configurable similar to other steps in pipeline, providing the ability to add custom algorithms for data preprocessing. 

Currently, the implementation is such that it will only use the available data preprocessing pipeline by default. If user wants to disable step, they can set include_data_preprocessor to ['no_preprocessing'] or can also use both available algos by setting it to ['no_preprocessing', 'feature_type']. It supports ability to add custom data preprocessing step. But I am facing issues handling init_params.  PR needs improvements.

Also, even data do not have any missing values, we have missing value imputation mandatory step. The output contains config for imputation strategy and gives false impression that imputation is performed even if there is no missing value in the data. For more enhancement, we can include enabling missing value imputation only when data has missing values otherwise forbid the configuration.

To summarise the goal:

- Ability to configure data preprocessing step in SimpleClassificationPipeline/SimpleRegressionPipeline (Implemented)
- Ability to configure each step in data preprocessing pipeline
- Ability to add new custom data preprocessing step (Implemented, missing proper handling of init_params)
- Enable steps in data preprocessing pipeline based on dataset properties

It will help if you can provide me more details regarding init_params and what all needs to be done to handle it properly. Also, how I can add ability to configure each step in data preprocessing pipeline.

**Comment by mfeurer** [comment]
> The idea of this PR is to make data preprocessing step configurable similar to other steps in pipeline, providing the ability to add custom algorithms for data preprocessing.

Got it.

> Currently, the implementation is such that it will only use the available data preprocessing pipeline by default. If user wants to disable step, they can set include_data_preprocessor to ['no_preprocessing'] or can also use both available algos by setting it to ['no_preprocessing', 'feature_type']. It supports ability to add custom data preprocessing step. But I am facing issues handling init_params. PR needs improvements.

I see. This totally makes sense, however, there's a small issue now in that 'no_preprocessing' is always an option now which can potentially lead to a lot of problems in case data has missing values. I guess it would be good if the no_preprocessing option would only be given in an example so it is not there by default.

> It will help if you can provide me more details regarding init_params and what all needs to be done to handle it properly. Also, how I can add ability to configure each step in data preprocessing pipeline.

Regarding the init_params, could you please give me some details on what is failing right now? Regarding configuring each step in data preprocessing, that'll be rather tough given the current API. I guess the way forward here would be to replace `include_estimators`, `include_preprocessors` and `include_data_preprocessor` by a single dictionary which allows addressing each component by name. I'll keep thinking about an API for that and will keep you posted.

As a final remark, please excuse the slow response time. I can only have a look at at most one or two PRs at a time, and currently your PR for multiple metrics is more important to me.

**Comment by rabsr** [comment]
>  there's a small issue now in that 'no_preprocessing' is always an option now which can potentially lead to a lot of problems in case data has missing values. I guess it would be good if the no_preprocessing option would only be given in an example so it is not there by default.

I agree. no_preprocessing should be handled carefully. This we can make sure by having more control over each sub component depending on data meta features, let's say only enable imputation in case of missing values in dataset. Currently, each subcomponent is enabled by default and uses its hyperparameters for optimization which may not be required. Also, it gives false impression that mean imputation is performed but data doesn't have any missing values. We can always raise exception if no_preprocessing is selected and data has missing values.

> Regarding the init_params, could you please give me some details on what is failing right now? 

[Here](https://github.com/automl/auto-sklearn/blob/development/autosklearn/evaluation/abstract_evaluator.py#L183), init_params is set for categorical features and if values are set are strictly checked by check_init_params method. This makes it somewhat difficult to make data_preprocessing configurable.

>Regarding configuring each step in data preprocessing, that'll be rather tough given the current API. I guess the way forward here would be to replace include_estimators, include_preprocessors and include_data_preprocessor by a single dictionary which allows addressing each component by name. I'll keep thinking about an API for that and will keep you posted.

Agree on that. Do you get any idea on how API should look like?

**Comment by mfeurer** [comment]
Sorry for the long round-trip time, I'm slowly catching up with older issues.

> Agree on that. Do you get any idea on how API should look like?

If you have a look at the way we currently construct the search space (https://github.com/automl/auto-sklearn/blob/2a4d388beef304c6fbdb86d89c80b5d85049931c/autosklearn/util/pipeline.py#L27), you'll see that's it's translated into a dictionary of a component name mapping to what to include and exclude for that component name, which is later on used here: https://github.com/automl/auto-sklearn/blob/2a4d388beef304c6fbdb86d89c80b5d85049931c/autosklearn/pipeline/base.py#L333

Having such an API where it would be possible to specify for each part of the pipeline what's in it would be a good first step moving towards a user being able to specify the complete pipeline theirselves.

**Comment by rabsr** [comment]
@mfeurer Thanks for reply. 

I checked how input is translated to dictionary of a component mapping for include and exclude list for component. I will make API changes to accept directly dictionary as an input. I will have include and exclude as two parameters keeping the same behaviour.

**Comment by rabsr** [comment]
@mfeurer I have updated the implementation with suggested changes and is ready for review. This PR:

- makes data_preprocessor step configurable, changing default implementation as `feature_type` component for the step
- adds example to demonstrate how to extend and skip data_preprocessor step
- updates configurations
- removes individual include/exclude parameter for each step. Components can now be included and excluded by passing dictionary with key as step name and value as list of components to be included/excluded, making sure that same step can not be present in both include and exclude.

File: autosklearn/estimators.py [hunk 1]
```diff
@@ -95,22 +93,14 @@ def __init__(
             In case of multi-processing, `memory_limit` will be per job.
             This memory limit also applies to the ensemble creation process.
 
-        include_estimators : list, optional (None)
-            If None, all possible estimators are used. Otherwise specifies
-            set of estimators to use.
+        include : dict, optional (None)
+            If None, all possible algorithms are used. Otherwise specifies
+            set of algorithms for each added component is used.
```
Comment (franchuterivera) [inline autosklearn/estimators.py:98]: Is it possible to add that include and exclude are incompatible if used together on the same component? For example, before it use to say:
```
Incompatible with include_estimators.
```

File: autosklearn/pipeline/base.py [hunk 2]
```diff
@@ -453,3 +455,10 @@ def get_additional_run_info(self):
         the optimization algorithm.
         """
         return self._additional_run_info
+
+    def _validate_include_exclude_params(self):
+        steps = [step[0] for step in self.steps]
+        for node in list(self.include.keys()) + list(self.exclude):
```
Comment (franchuterivera) [inline autosklearn/pipeline/base.py:461]: Shouldn't this be:
```
for node in list(self.include.keys()) + list(self.exclude.keys()):
```

File: autosklearn/pipeline/components/data_preprocessing/feature_type.py [hunk 3]
```diff
@@ -40,31 +41,31 @@ def __init__(self, config=None, pipeline=None, dataset_properties=None, include=
...
         # the categorical pipeline. Hence, force to None
         # It is actually the call to set_hyperparameter who properly sets this argument
-        # TODO: Extract the child configuration space from the datapreprocessor to the
+        # TODO: Extract the child configuration space from the FeatTypeSplit to the
         # pipeline if needed
         self.categ_ppl = CategoricalPreprocessingPipeline(
             config=None, steps=pipeline, dataset_properties=dataset_properties,
-            include=include, exclude=exclude, random_state=random_state,
-            init_params=init_params)
+            random_state=random_state, init_params=init_params)
```
Comment (franchuterivera) [inline autosklearn/pipeline/components/data_preprocessing/feature_type.py:58]: Sorry for the trouble, could you elaborate why include and exclude are removed here?

File: autosklearn/smbo.py [hunk 4]
```diff
@@ -411,41 +406,19 @@ def run_smbo(self):
...
-            else:
-                raise ValueError(self.task)
-        elif self.exclude_estimators is not None:
-            if self.task in CLASSIFICATION_TASKS:
-                exclude['classifier'] = self.exclude_estimators
-            elif self.task in REGRESSION_TASKS:
-                exclude['regressor'] = self.exclude_estimators
-            else:
-                raise ValueError(self.task)
+        if self.include is not None and self.exclude is not None:
```
Comment (franchuterivera) [inline autosklearn/smbo.py:409]: I think that if you rebase the code to the latest version, this changed a bit.

**Comment by franchuterivera** [comment]
Hello, thanks a lot for the PR and please excuse the delay. Things look great, and if you could rebase your changes to the development branch, we can double-check that all tests pass. I did leave some minor questions.

I think these changes make a lot of sense and add flexibility for the user to interact with the pipeline steps. It is an advanced feature, as the user might break the pipeline if no preprocessing existed at all on a dataset that requires it. Another possibility would have been to change the column transformer, but it would not have been an isolated change like this one.

@mfeurer do you have a suggestion to check the metalearning files?

File: autosklearn/estimators.py (hunk 1 above)
Comment (rabsr) [reply autosklearn/estimators.py]: Done.

File: autosklearn/pipeline/base.py (hunk 2 above)
Comment (rabsr) [reply autosklearn/pipeline/base.py]: Thanks for pointing it. Made changes.

File: autosklearn/smbo.py (hunk 4 above)
Comment (rabsr) [reply autosklearn/smbo.py:414]: Rebased. Looks same to me. There are no changes over here

File: autosklearn/pipeline/components/data_preprocessing/feature_type.py (hunk 3 above)
Comment (rabsr) [reply autosklearn/pipeline/components/data_preprocessing/feature_type.py]: Earlier there no configuration for data preprocessing step. Thus None was used for include/exclude here. But as now the step is configurable, we need to carefully handle the step. It's kinda parallel pipelines (which have configurable steps such as rescaling, cat encoding) within pipeline. For now I added it back, and removed include and exclude from where they were passed here. 
Please let me know if it is required to have include and exclude for num/cat pipeline steps. We may need to think how the schema of include/exclude option look like for complex pipeline structures. 
Example for current implemented schema for include/exclude
```
{
    'classifier' : ['sgd','lda'],
    'feature_preprocessor': ['pca', 'kernel_pca']
}
```
What do you think of below schema structure for complex pipelines?
```
{
        'classifier': ['sgd', 'lda'],
        'feature_preprocessor': ['pca', 'kernel_pca'],
        'data_preprocessor': {
            'numerical_transformer': {
                'rescaling': ['minmax', 'normalize']
            },
            'categorical_transformer': {
                'category_coalescence': ['no_coalescense'],
                'categorical_encoding': ['no_encoding']
            }
        }
    }
```
Please let me know if this makes sense I will make changes accordingly.

**Comment by rabsr** [comment]
@franchuterivera Thanks for the feedback. I have rebased and made changes as suggested. I have replied back with potential enhancements on one of the review comments. Please let know what are your thoughts on this.

I have one more enhancement to discuss which is as follows:
Move preprocessors such as variance threshold, category shift and category encoding to input validator. As these are constant steps and no tuning happens on these steps, they can be performed initially just once on the data like categorical label encoding. This will help in saving some time per pipeline and possibly more config tries in total time.

File: autosklearn/pipeline/components/data_preprocessing/feature_type.py (hunk 3 above)
Comment (franchuterivera) [reply autosklearn/pipeline/components/data_preprocessing/feature_type.py]: Thanks for your reply. So the question we want to answer here is if someone wants to use for instance only `'minmax'` scaling. But what happens if there are no numerical columns? We have to take into account multiple scenarios.

To my understanding, if someone wants to use only `minmax` as a data-preprocessing he/she can do so via:
```
autosklearn.pipeline.components.data_preprocessing.add_preprocessor(<custom preprocessing>)
```

So adding support for these custom scenarios might complicate the code and there is a work-around solution for these custom cases. In order words, I think your change gives enough flexibility to the users as it is. What do you think?

File: autosklearn/smbo.py (hunk 4 above)
Comment (franchuterivera) [reply autosklearn/smbo.py:414]: Yes, sorry. I thought a change was already merged (#1096). Depending on what gets merged first, we will update the code accordingly

**Comment by franchuterivera** [comment]
> Move preprocessors such as variance threshold, category shift and category encoding to input validator. As these are constant steps and no tuning happens on these steps, they can be performed initially just once on the data like categorical label encoding. This will help in saving some time per pipeline and possibly more config tries in total time.

Maybe this makes sense on another PR, but we are also trying to change the way input validator is used today.  It will be great for example, that each pipeline can predict in a self-sufficient manner. Today the pipeline is dependent on the main estimator due to the input validator, and maybe a user just wants to have a single pipeline as the final solution. We are still exploring solutions.

File: autosklearn/pipeline/components/data_preprocessing/feature_type.py (hunk 3 above)
Comment (rabsr) [reply autosklearn/pipeline/components/data_preprocessing/feature_type.py]: Agreed, it will complicate code. We can go ahead with current implementation. Let me know if you have any other feedback.

File: autosklearn/smbo.py (hunk 4 above)
Comment (rabsr) [reply autosklearn/smbo.py:414]: Sure.

**Comment by rabsr** [comment]
@franchuterivera Thanks for the update. Having single pipeline as the final solution makes much sense. I am looking forward to it.

Adding the script which is used to update metalearning configuration files. Hope this helps.
```
import csv
import glob

WINDOWS_LINE_ENDING = b'\r\n'
UNIX_LINE_ENDING = b'\n'


def update_config(file_path):
    config_file = open(file_path, "r")
    config = csv.DictReader(config_file)
    field = config.fieldnames.copy()
    field.append("data_preprocessor:__choice__")
    rows = []
    for row in config:
        row["data_preprocessor:__choice__"] = "feature_type"
        rows.append(row)

    new_config_file = open(file_path, "w")
    new_config = csv.DictWriter(new_config_file, fieldnames=field)
    new_config.writeheader()
    new_config.writerows(rows)
    new_config_file.close()

    with open(file_path, "rb") as f:
        new_con = f.read()
        new_con = new_con.replace(b"data_preprocessing", b"data_preprocessor:feature_type") \
            .replace(b"\0", b"") \
            .replace(WINDOWS_LINE_ENDING, UNIX_LINE_ENDING)

    with open(file_path, "wb") as f:
        f.write(new_con)


files = glob.glob(
    "autosklearn/metalearning/files/**/configurations.csv")
for file in files:
    try:
        update_config(file)
    except:
        print(file)
        raise
```

File: autosklearn/smbo.py (hunk 4 above)
Comment (rabsr) [reply autosklearn/smbo.py:412]: Pulled latest changes after #1096 merge.

**Comment by franchuterivera** [comment]
Hi @rabsr I will be checking the metadata and get back to you shortly (hopefully no longer than a day or two).

I discussed with Matthias offline and we think there is room for improvement in [this method](https://github.com/automl/auto-sklearn/blob/47aed90e5bacbf820d3ea71a11ddc2d8fe40a575/autosklearn/pipeline/base.py#L459)

Assuming we will have a new user the new syntax might be complicated. So do you think we can improve the error checking here? This means:

- Check if the format of the include is ok. Currently, the keys of the dict are checked. Do you think we can also check that the values are list and provide a meaningful error to the user if that is not the case?
- Improve the error message in case we have a wrong key provided. For example say: `The provided key in the include argument is not valid. The only supported keys for classification are {}.` 
- Do you think we can be robust against extra keys?

These checks do not have to be where I indicated, if you can think of a better place, then that's even better. 

We just want to make sure Auto-Sklearn is not harder for new people :)

**Comment by rabsr** [comment]
@mfeurer Thanks for the feedback. I totally agree that it should be easy to use for everyone and the errors must be self explanatory. I have added more checks to validate the structure and values of the include/exclude arguments as follows:
- Both include and exclude cant have same steps
- No extra key will be accepted in the include/exclude dict. The keys are validated against list of supported steps(steps which involves algorithm selection)
- The values of each key must be non empty list
- The values of each key is validated against supported components for the step

I have also improved on the error message.

File: autosklearn/pipeline/base.py [hunk 5]
```diff
@@ -19,13 +19,15 @@ class BasePipeline(Pipeline):
 
     def __init__(self, config=None, steps=None, dataset_properties=None,
                  include=None, exclude=None, random_state=None,
-                 init_params=None):
+                 init_params=None, target_type=None):
```
Comment (franchuterivera) [inline autosklearn/pipeline/base.py:22]: Hello, thanks a lot for the enhancements, they look pretty good.

This is the only change I am not fully able to understand. Do you think we can instead of defining another target_type, get it on the dataset properties?

For example, we define that [here](https://github.com/automl/auto-sklearn/blob/275d0d6b20d16822252d8b50bf71b1c787187f09/autosklearn/pipeline/classification.py#L161). I am wondering in which scenarios is the `target_type` missing?

Comment (rabsr) [reply autosklearn/pipeline/base.py:22]: In recent changes, to validate include/exclude params exhaustively, we are also validating for the components for the steps in these params. To validate against all [available components](https://github.com/rabsr/auto-sklearn/blob/config_data_preprocess/autosklearn/pipeline/base.py#L486), `target_type` is required in `dataset_properties` for [feature preprocessors](https://github.com/rabsr/auto-sklearn/blob/config_data_preprocess/autosklearn/pipeline/components/feature_preprocessing/__init__.py#L62)  and [data preprocessors](https://github.com/rabsr/auto-sklearn/blob/config_data_preprocess/autosklearn/pipeline/components/data_preprocessing/__init__.py#L63) 

Now `target_type` is required in `dataset_properties` while creating pipeline objects and it is missing at all places ([utils/pipeline.py](https://github.com/rabsr/auto-sklearn/blob/config_data_preprocess/autosklearn/util/pipeline.py#L82) as well as [abstract_evaluator.py](https://github.com/rabsr/auto-sklearn/blob/config_data_preprocess/autosklearn/evaluation/abstract_evaluator.py#L284)). Instead of adding it at various places and then validating and updating when creating pipeline object, I tried to have minimum changes to achieve the same.

Please do let me know if you have better approach to handle this. I will improve on implementation.

Comment (franchuterivera) [reply autosklearn/pipeline/base.py:22]: Thanks a lot for the explanation.

Yeah, this is clearly needed when you instantiate the object, and in particular when dataset_properties is `None`/missing this target_type.

I do have one implementation change request to achieve the same goal. Can you add this change [here](https://github.com/automl/auto-sklearn/blob/81115f6d1184966c872a82403a3c6abc60cef9db/autosklearn/pipeline/classification.py#L74) rather than a new attribute?

Something like this:
```
if dataset_properties is None or 'target_type' not in dataset_properties:
    --> then add it here in the classification/regression pipeline
```

Why not an argument to the pipeline? This complicates things with scikit learn, for example with `get_params`/`set_params` you need all attributes to be stored as `self.ATTRIBUTE_NAME`. Also, we are trying to keep the interface the same if possible.

Comment (rabsr) [reply autosklearn/pipeline/base.py]: Thanks for clarifying. Have updated as per your suggestion.

**Comment by rabsr** [comment]
@mfeurer @franchuterivera Can anyone of you help with the test cases, what's this error about?

**Comment by mfeurer** [comment]
The error message is about this patch breaking the parallel mode of Auto-sklearn. However, the error actually also happens in sequential mode. The difference is just that parallel breaks the unit tests. Therefore, I suggest that you run the first failing test on your local machine to see what the issue is.

**Comment by rabsr** [comment]
@mfeurer I have fixed all the test cases and rebased to development branch. Please check now.

**Comment by eddiebergman** [comment]
Hi @rabsr,

It looks good to me. Unfortunately I missed out on most of this discussion as it was happening, and the rebase has killed the file changes to show all of the development changes so my ability to do my own review is hampered.

Seeing as @mfeurer and @franchuterivera were happy up to commit #12e56e3d804120b4a19939c2400c96be523f83c6 and all changes since were test and build fixes, I am happy to merge it as it's been on the radar for a while.

If any issues are to appear down the line, I will create an issue and tag you in!

Thanks again and sorry for the slow response time :)
